src/main/java/bio/terra/common/CloudPlatformWrapper.java

```diff
@@ -0,0 +1,126 @@
+
+    public abstract GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel datasetRequestModel);
+
+    static class GcpPlatform extends CloudPlatformWrapper {
+        static final GcpPlatform INSTANCE = new GcpPlatform();
+
+        @Override
+        public boolean isGcp() {
+            return true;
+        }
+
+        @Override
+        public void ensureValidRegion(String region, Errors errors) {
+            ensureValidRegion(CloudPlatform.GCP, region, GoogleRegion.SUPPORTED_REGIONS, errors);
+        }
+
+        @Override
+        public List<? extends StorageResource<?, ?>> createStorageResourceValues(DatasetRequestModel datasetRequest) {
+            final GoogleRegion region = getRegionFromDatasetRequestModel(datasetRequest);
+            return Arrays.stream(GoogleCloudResource.values()).map(resource -> {
+                final GoogleRegion finalRegion;
+                switch (resource) {
+                case FIRESTORE: finalRegion = region.getRegionOrFallbackFirestoreRegion();
+                    break;
+                case BUCKET: finalRegion = region.getRegionOrFallbackBucketRegion();
+                    break;
+                default: finalRegion = region;
+                }
+                return new GoogleStorageResource(null, resource, finalRegion);
+            }).collect(Collectors.toList());
+        }
+
+        @Override
+        public GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel datasetRequestModel) {
+            return GoogleRegion.fromValueWithDefault(datasetRequestModel.getRegion());
+        }
+    }
+
+    static class AzurePlatform extends CloudPlatformWrapper {
+        static final AzurePlatform INSTANCE = new AzurePlatform();
+
+        @Override
+        public boolean isAzure() {
+            return true;
+        }
+
+        @Override
+        public void ensureValidRegion(String region, Errors errors) {
+            ensureValidRegion(CloudPlatform.AZURE, region, AzureRegion.SUPPORTED_REGIONS, errors);
+            ensureValidRegion(CloudPlatform.GCP, region, GoogleRegion.SUPPORTED_REGIONS, errors);
+        }
+
+        @Override
+        public List<? extends StorageResource<?, ?>> createStorageResourceValues(DatasetRequestModel datasetRequest) {
+            final AzureRegion region = AzureRegion.fromValueWithDefault(datasetRequest.getRegion());
+            // TODO: once we no longer require GCP resources to back Azure datasets, stop concatenating
+            return Stream.concat(
+                Arrays.stream(AzureCloudResource.values()).map(resource -> new AzureStorageResource(null,
+                    resource,
+                    region)),
+                CloudPlatformWrapper.of(CloudPlatform.GCP).createStorageResourceValues(datasetRequest).stream())
+                .collect(Collectors.toList());
+        }
+
+        @Override
+        public GoogleRegion getRegionFromDatasetRequestModel(DatasetRequestModel datasetRequestModel) {
```

Comment:
But this needs to be an azure region...and we've already built out this abstraction in the CloudRegion class...

Reply:
Maybe it should be `AzureRegion`, but it wasn't in the original code. This is copied from `DatasetJsonConversion` below. https://github.com/DataBiosphere/jade-data-repo/pull/984/files#diff-58fdba484011b4cdb292414f5a7c3aba295019f828cfb083c0c21bfca5af90ddL88